src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs

```diff
@@ -229,12 +317,35 @@ private class TokenResolverProvider : ISignatureTypeProvider<DummyTypeInfo, Modu
 
             IEcmaModule _contextModule;
 
+            [ThreadStatic]
+            private static TokenResolverProvider _rentalObject;
+
             public TokenResolverProvider(ModuleTokenResolver resolver, IEcmaModule contextModule)
             {
                 _resolver = resolver;
                 _contextModule = contextModule;
             }
 
+            public static TokenResolverProvider Rent(ModuleTokenResolver resolver, IEcmaModule contextModule)
```

Comment:
Just curious - wouldn't it be even more performant to use a <code>ConcurrentDictionary&lt;IEcmaModule, TokenResolverProvider&gt;</code> provided we flush it after each compilation? While the <code>TokenResolverProvider</code> rental is a smart micro-optimization, it doesn't prevent repeated provider creation for the same context module which seems superfluous - does it mean that the occasional repeated allocation is still cheaper than using a dictionary for unification?

Reply:
Eh, this is probably cheaper, as we almost never actually need to allocate a new one, and TLS lookup is really quite fast.